module/state_synchronization/indexer/indexer_core.go

```diff
@@ -0,0 +1,178 @@
+func (i *IndexerCore) RegisterValues(IDs flow.RegisterIDs, height uint64) ([]flow.RegisterValue, error) {
+	values := make([]flow.RegisterValue, len(IDs))
+
+	for j, id := range IDs {
+		value, err := i.registers.Get(id, height)
+		if err != nil {
+			return nil, err
+		}
+
+		values[j] = value
+	}
+
+	return values, nil
+}
+
+// IndexBlockData indexes all execution block data by height.
+// This method shouldn't be used concurrently.
+// Expected errors:
+// - ErrIndexValue if height was not incremented in sequence
+// - storage.ErrNotFound if the block for execution data was not found
+func (i *IndexerCore) IndexBlockData(_ context.Context, data *execution_data.BlockExecutionDataEntity) error {
+	fmt.Println("index block data", data.BlockID)
+	block, err := i.headers.ByBlockID(data.BlockID)
+	if err != nil {
+		return fmt.Errorf("could not get the block by ID %s: %w", data.BlockID, err)
+	}
+
+	lg := i.log.With().
+		Hex("block_id", logging.ID(data.BlockID)).
+		Uint64("height", block.Height).
+		Logger()
+
+	lg.Debug().Msgf("indexing new block")
+
+	// the height we are indexing must be exactly one bigger or same as the latest height indexed from the storage
+	latest := i.registers.LatestHeight()
+	if block.Height != latest+1 && block.Height != latest {
+		return fmt.Errorf("must store registers with the next height %d, but got %d", latest+1, block.Height)
+	}
+	// allow rerunning the indexer for same height since we are fetching height from register storage, but there are other storages
+	// for indexing resources which might fail to update the values, so this enables rerunning and reindexing those resources
+	if block.Height == latest {
+		lg.Warn().Msg("reindexing block data")
+	}
+
+	// concurrently process indexing of block data
+	g := errgroup.Group{}
+
+	updates := make([]*ledger.TrieUpdate, 0)
+	events := make([]flow.Event, 0)
+
+	for _, chunk := range data.ChunkExecutionDatas {
+		if chunk.TrieUpdate != nil {
+			updates = append(updates, chunk.TrieUpdate)
+		}
+
+		events = append(events, chunk.Events...)
+	}
+
+	if len(events) > 0 {
+		g.Go(func() error {
+			err := i.indexEvents(data.BlockID, events)
+			if err != nil {
+				return fmt.Errorf("could not index events at height %d: %w", block.Height, err)
+			}
+			return nil
+		})
+	}
+
+	g.Go(func() error {
+		// we are iterating all the registers and overwrite any existing register at the same path
+		// this will make sure if we have multiple register changes only the last change will get persisted
+		// if block has two chucks:
+		// first chunk updates: { X: 1, Y: 2 }
+		// second chunk updates: { X: 2 }
+		// then we should persist only {X: 2: Y: 2}
+
+		payloads := make(map[ledger.Path]*ledger.Payload)
+		for _, update := range updates {
+			for i, path := range update.Paths {
+				payloads[path] = update.Payloads[i] // todo should we use TrieUpdate.Paths or TrieUpdate.Payload.Key?
+			}
+		}
+
+		err = i.indexRegisterPayloads(maps.Values(payloads), block.Height)
```

Comment:
`maps.Values` can be removed, since we've deduplicated registers, we can just pass the map, and iterate the map to create a list of registers to be stored

Reply:
I just used this method instead of writing a for loop to create a list. Wouldn't that be just extra code or is this call more expensive and that's why?

Reply:
Ohh I see what you mean. Good point.